### PR TITLE
Connec! external id, smart merging and big refactoring

### PR DESCRIPTION
**Pull request by Berardpi**
@BrunoChauvet @x4d3 
A bit more change than I originally expected, but I figured it was now or never.

PR may be messy to review as there's a lot of method and parameter changes. You can have a look at the `release_note.md` file first.

I've added integration specs to minimize the risk of major bugs and not cause too many issues on production released connectors. (99.87% coverage 😞 )

If you have anything that you think may be improve don't hesitate, now is the time.

I'll migrate salesforce and constant contact (open sourced ones first) to test.

### CODE REVIEW

**Comment by Berardpi** [comment]
I've migrated SF to test (https://github.com/maestrano/connector-salesforce/pull/43) if you want to see what the diff looks like. Lead to minor fixes and improvements.

File: app/controllers/maestrano/connec_controller.rb [hunk 1]
```diff
@@ -6,11 +6,9 @@ def notifications
     begin
       params.except(:tenant, :controller, :action, :connec).each do |entity_name, entities|
 
-        entity_instance_hash = find_entity_instance(entity_name)
+        entity_instance_hash = find_entity_class(entity_name)
```
Comment (x4d3) [inline app/controllers/maestrano/connec_controller.rb]: rename to entity_class_hash

File: app/jobs/maestrano/connector/rails/push_to_connec_job.rb [hunk 2]
```diff
@@ -11,35 +11,35 @@ def perform(organization, entities_hash, opts={})
...
           next unless organization.synchronized_entities[entity_instance_hash[:name].to_sym]
 
           entity_instance = entity_instance_hash[:instance]
 
-          entity_instance.before_sync(connec_client, external_client, last_synchronization, organization, opts)
+          entity_instance.before_sync(last_synchronization)
           # Build expected input for consolidate_and_map_data
           if entity_instance_hash[:is_complex]
-            mapped_entities = entity_instance.consolidate_and_map_data(Hash[ *entity_instance.class.connec_entities_names.collect{|name| [ name, []]}.flatten(1) ], Hash[ *entity_instance.class.external_entities_names.collect{|name| name == external_entity_name ? [name, entities] : [ name, []]}.flatten(1) ], organization, {})
+            mapped_entities = entity_instance.consolidate_and_map_data(Hash[ *entity_instance.class.connec_entities_names.collect{|name| [ name, []]}.flatten(1) ], Hash[ *entity_instance.class.external_entities_names.collect{|name| name == external_entity_name ? [name, entities] : [ name, []]}.flatten(1) ])
```
Comment (x4d3) [inline app/jobs/maestrano/connector/rails/push_to_connec_job.rb]: please add some comment on what this lovely line is doing, or extract it to a private static method that describe the operation.

File: app/controllers/maestrano/connec_controller.rb [hunk 3]
```diff
@@ -22,20 +20,20 @@ def notifications
...
-          entity_instance.before_sync(connec_client, external_client, last_synchronization, organization, {})
+          entity_instance = entity_instance_hash[:class].new(organization, connec_client, external_client, {})
+          entity_instance.before_sync(last_synchronization)
           
-
           # Build expected input for consolidate_and_map_data
           if entity_instance_hash[:is_complex]
-            filtered_entities = entity_instance.filter_connec_entities(Hash[ *entity_instance.class.connec_entities_names.collect{|name| name.parameterize('_').pluralize == entity_name ? [name, [entity]] : [ name, []]}.flatten(1) ], organization)
-            mapped_entity = entity_instance.consolidate_and_map_data(filtered_entities, Hash[ *entity_instance.class.external_entities_names.collect{|name| [ name, []]}.flatten(1) ], organization, {})
+            filtered_entities = entity_instance.filter_connec_entities(Hash[ *entity_instance.class.connec_entities_names.collect{|name| name.parameterize('_').pluralize == entity_name ? [name, [entity]] : [ name, []]}.flatten(1) ])
```
Comment (x4d3) [inline app/controllers/maestrano/connec_controller.rb]: please could you write some comment on this cute line ?

File: app/jobs/maestrano/connector/rails/synchronization_job.rb [hunk 4]
```diff
@@ -54,15 +54,15 @@ def perform(organization, opts)
...
-      mapped_entities = entity_instance.consolidate_and_map_data(connec_entities, external_entities, organization, opts)
-      entity_instance.push_entities_to_external(external_client, mapped_entities[:connec_entities], organization)
-      entity_instance.push_entities_to_connec(connec_client, mapped_entities[:external_entities], organization)
-      entity_instance.after_sync(connec_client, external_client, last_synchronization, organization, opts)
+      entity_instance.before_sync(last_synchronization)
+      external_entities = entity_instance.get_external_entities(last_synchronization)
+      connec_entities = entity_instance.get_connec_entities(last_synchronization)
+      mapped_entities = entity_instance.consolidate_and_map_data(connec_entities, external_entities)
+      entity_instance.push_entities_to_external(mapped_entities[:connec_entities])
+      entity_instance.push_entities_to_connec(mapped_entities[:external_entities])
```
Comment (x4d3) [inline app/jobs/maestrano/connector/rails/synchronization_job.rb]: love it !

File: app/models/maestrano/connector/rails/concerns/complex_entity.rb [hunk 5]
```diff
@@ -45,100 +52,65 @@ def connec_model_to_external_model(connec_hash_of_entities, organization)
...
+  end
 
-          # Entity has not been modified since its last push to connec!
-          next nil if Maestrano::Connector::Rails::Entity.not_modified_since_last_push_to_connec?(idmap, entity, sub_entity_instance, organization)
+  def consolidate_and_map_connec_entities(modeled_connec_entities, modeled_external_entities)
+    modeled_connec_entities.each do |connec_entity_name, entities_in_external_model|
+      entities_in_external_model.each do |external_entity_name, entities|
 
-          idmap.update(name: sub_entity_instance.class.object_name_from_external_entity_hash(entity))
+        sub_entity_instance = "Entities::SubEntities::#{connec_entity_name.titleize.split.join}".constantize.new(@organization, @connec_client, @external_client, @opts)
```
Comment (x4d3) [inline app/models/maestrano/connector/rails/concerns/complex_entity.rb]: extract to a method ? (saw it twice...)

File: app/models/maestrano/connector/rails/concerns/connec_helper.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
+module Maestrano::Connector::Rails::Concerns::ConnecHelper
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    
+    # Replace the ids arrays by the external id
+    # If a reference has no id for this oauth_provider and oauth_uid but has one for connec returns nil
+    def unfold_references(connec_entity, references, organization)
+      connec_entity = connec_entity.with_indifferent_access
```
Comment (x4d3) [inline app/models/maestrano/connector/rails/concerns/connec_helper.rb]: I would personnaly introduce a new variable name, like result, or unfold_connec_entities, to make it clear that there is not side effect on the connec_entity entry variable (personnal taste, I love obviously "pure method")

File: app/models/maestrano/connector/rails/concerns/sub_entity_base.rb [hunk 7]
```diff
@@ -34,72 +34,43 @@ def names_hash
...
-        field = self.class.external? ? ref[:connec_field] : ref[:external_field]
-        ref_hash.merge! field.split('/').reverse.inject(Maestrano::Connector::Rails::Entity.id_from_ref(entity, ref, self.class.external?, organization)) { |a, n| { n.to_sym => a } }
-      end
-    end
-
     if self.class.external?
-      mapped_entity = mapper.denormalize(entity)
+      mapped_entity = mapper.denormalize(entity).merge(id: self.class.id_from_external_entity_hash(entity))
+      folded_entity = Maestrano::Connector::Rails::ConnecHelper.fold_references(mapped_entity, self.class.references[name] || [], @organization)
+      folded_entity.merge!(opts: (mapped_entity[:opts] || {}).merge(matching_fields: self.class.connec_matching_fields)) if self.class.connec_matching_fields
```
Comment (x4d3) [inline app/models/maestrano/connector/rails/concerns/sub_entity_base.rb]: Here, if self.class.connec_matching_fields is nil, folded_entity won't be merge to opts: (mapped_entity[:opts] || {}, 

Is it wanted ?

File: release_notes.md [hunk 8]
```diff
@@ -1,3 +1,61 @@
+## 1.0.0
```
Comment (x4d3) [inline release_notes.md]: Awesome initiative.

Steak sauce!

**Comment by x4d3** [comment]
Except my tiny remarks, looks good! I tip my :tophat: .

**Comment by Berardpi** [comment]
Reverted depreciation of connec_id

**Pull request closed without merging**